terraform/eval_validate.go

```diff
@@ -264,12 +260,13 @@ type EvalValidateResource struct {
 	ConfigVal *cty.Value
 }
 
-func (n *EvalValidateResource) Validate(ctx EvalContext) error {
+func (n *EvalValidateResource) Validate(ctx EvalContext) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
 	if n.ProviderSchema == nil || *n.ProviderSchema == nil {
-		return fmt.Errorf("EvalValidateResource has nil schema for %s", n.Addr)
+		diags = diags.Append(fmt.Errorf("EvalValidateResource has nil schema for %s", n.Addr))
 	}
 
-	var diags tfdiags.Diagnostics
 	provider := *n.Provider
 	cfg := *n.Config
 	schema := *n.ProviderSchema
```

Comment:
Since we're no longer early returning, there looks like a _technical_ possibility of `n.ProviderSchema == nil` being true and this line panicking.

I tried untangling whether or not this is possible, and while I'm not sure it looks like it might be: [`NodeValidatableResource`](https://github.com/hashicorp/terraform/blob/fd43bc7847c01caa72ca9fec7c19cfb738d4b3a1/terraform/node_resource_validate.go#L43) -> [`GetProvider`](https://github.com/hashicorp/terraform/blob/fd43bc7847c01caa72ca9fec7c19cfb738d4b3a1/terraform/eval_provider.go#L60) -> [`BuiltinEvalContext`](https://github.com/hashicorp/terraform/blob/fd43bc7847c01caa72ca9fec7c19cfb738d4b3a1/terraform/eval_context_builtin.go#L153) ->  [`ProviderSchema`](https://github.com/hashicorp/terraform/blob/fd43bc7847c01caa72ca9fec7c19cfb738d4b3a1/terraform/schemas.go#L29-L31).

Would we lose anything useful if we early-return on nil schema?

Reply:
Good catch! I was just accumulating diagnostics here and did not double check that.
Looking more closely, neither `n.ProviderSchema` nor `*n.ProviderSchema` should ever be nil (the former is more easily verified to never be nil than the latter however). While nothing would work if `GetProvider` didn't return a schema, we can probably make this a early return again just to be safe.